lib/sigobj.js

```diff
@@ -0,0 +1,254 @@
+/*!
+ * sigobj.js - Signature object
+ * Copyright (c) 2018, The Bcoin Developers (MIT License).
+ * https://github.com/bcoin-org/bcoin
+ */
+
+'use strict';
+
+const assert = require('assert');
+const MTX = require('bcoin/lib/primitives/mtx');
+const LedgerBTCApp = require('./ledger');
+const secp256k1 = require('bcrypto').secp256k1;
+
+/**
+ * Wraps and provides all
+ * necessary information for
+ * signing.
+ */
+
+class SignatureObject {
+  /**
+   * Create signature object
+   * @param {Object} options
+   * @param {MTX|JSON} options.mtx
+   * @param {LedgerTXInput[]} options.inputs
+   */
+
+  constructor(options) {
+    this.ledger = null;
+    this.mtx = new MTX();
+    this.inputs = [];
+    this.inputsByKey = new Map();
+    this.indexByInput = new Map();
+    this.trustedInputs = new Map();
+    this.witness = false;
+    this.new = true;
+
+    this.signedInputs = 0;
+
+    // process
+    this.initialized = false;
+
+    // cache some info
+    this._trustedInputs = false;
+    this._publicKeys = false;
+
+    if (options)
+      this.fromOptions(options);
+  }
+
+  /**
+   * Inject properties from options
+   * @param {Object} options
+   * @returns {SignatureObject}
+   */
+
+  fromOptions(options) {
+    assert(options, 'SignatureObject options are required.');
+    assert(options.mtx, 'MTX is required.');
+    assert(options.ledger, 'Ledger app is required.');
+    assert(options.ledger instanceof LedgerBTCApp,
+      'Ledger is not LedgerBTCApp instance.');
+
+    this.ledger = options.ledger;
+
+    if (!MTX.isMTX(options.mtx) && typeof options.mtx === 'object')
+      options.mtx = MTX.fromJSON(options.mtx);
+
+    if (MTX.isMTX(options.mtx))
+      this.mtx = options.mtx;
+
+    if (options.inputs) {
+      assert(Array.isArray(options.inputs));
+      this.inputs = options.inputs;
+    }
+
+    return this;
+  }
+
+  /**
+   * Create sigobject from options
+   * @param {Object} options
+   * @returns {SignatureObject}
+   */
+
+  static fromOptions(options) {
+    return new this(options);
+  }
+
+  /**
+   * Initialize maps with inputs
+   */
+
+  init() {
+    assert(this.ledger);
+    assert(!this.initialized);
+
+    for (const li of this.inputs) {
+      this.inputsByKey.set(li.toKey(), li);
+
+      if (li.witness)
```

Comment:
Is there any issue if not all inputs are either witness `true` or `false`?

Reply:
Nope, but they are processed differently if any of them is